spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,549 @@
+
+# %% [markdown]
+# For our $n$ examples and $k$ slices in each split, we apply the SF to our data to create an $n \times k$ matrix. (So far, $k=1$).
+
+# %%
+from snorkel.slicing.apply import PandasSFApplier
+
+applier = PandasSFApplier(sfs)
+S_train = applier.apply(df_train)
+S_valid = applier.apply(df_valid)
+S_test = applier.apply(df_test)
+
+# %% [markdown]
+# With a utility function from `snorkel.analysis.monitor`, we can visualize examples belonging to this slice in a `pandas.DataFrame`.
+
+# %%
+from snorkel.analysis.monitor import slice_df
+
+short_link_df = slice_df(
+    S_matrix=S_train, slice_names=slice_names, target_slice="short_link", df=df_train
+)
+
+short_link_df[["text", "label"]]
+
+# %%
+from snorkel.analysis.monitor import slice_df
+
+short_link_df = slice_df(
+    S_matrix=S_valid, slice_names=slice_names, target_slice="short_link", df=df_valid
+)
+
+short_link_df[["text", "label"]]
+
+# %% [markdown]
+# Now, we add labels for this particularly slice to an existing dataloader.
+# Specifically, `add_slice_labels` will add two sets of labels for each slice:
+# * `spam_task_slice:{slice_name}_ind`: an indicator label, which corresponds to the outputs of the slicing functions.
+# These indicate whether each example is in the slice (`label=1`)or not (`label=0`).
+# * `spam_task_slice:{slice_name}_pred`: a _masked_ set of the original task labels (in this case, labeled `spam_task`) for each slice. Examples that are masked (with `label=-1`) will not contribute to loss or scoring.
+
+# %%
+from snorkel.slicing.utils import add_slice_labels
+
+slice_names = [sf.name for sf in sfs]
+add_slice_labels(dl_train, spam_task, S_train, slice_names)
+add_slice_labels(dl_valid, spam_task, S_valid, slice_names)
+add_slice_labels(dl_test, spam_task, S_test, slice_names)
+
+# %%
+dl_valid.dataset
+
+# %% [markdown]
+# With our updated dataloader, we want to evaluate on model on the defined slice. In the `SnorkelClassifier`, we can call `score` with an additional argument, `remap_labels` to specify that the slice's prediction labels, `spam_task_slice:short_link_pred` should be mapped to the `spam_task` for evaluation.
+
+# %%
+model.score(
+    dataloaders=[dl_valid, dl_test],
+    remap_labels={"spam_task_slice:short_link_pred": "spam_task"},
+    as_dataframe=True,
+)
+
+# %% [markdown]
+# ### Monitor slices in non-Snorkel models
+
+# %% [markdown]
+# If you're using a model other than `SnorkelClassifier`, you can still evaluate on slices using the `score_on_slices` helper.
+#
+# We define a `LogisticRegression` model from sklearn and show how we might visualize these slice-specific scores.
+
+# %%
+from sklearn.linear_model import LogisticRegression
+
+sklearn_model = LogisticRegression(C=0.001, solver="liblinear")
+sklearn_model.fit(X=X_train, y=Y_train)
+sklearn_model.score(X_test, Y_test)
+
+# %%
+from snorkel.analysis.monitor import score_on_slices
+
+test_preds = sklearn_model.predict(X_test)
+
+score_on_slices(
```

Comment:
would be nice to not have to generate S_test separately (in this flow, there's no other use for it)

so 

```python
slice_scorer = PandasSliceScorer(sfs, metrics)
slice_scorer.score(df_test, Y_test, test_preds)
```

slice_scorer can be reused for multiple dataset splits

Reply:
why would we require all there `(df_test, Y_test, test_preds)` as args?

Reply:
re: in-person convo 

```
slice_scorer = SliceScorer(S_matrix, slice_names, metrics)
slice_scorer.score(golds, preds)
```